Alarm based timeouts can only be used in the main thread.

Observed this failure in [this GH run](https://github.com/PrefectHQ/prefect/actions/runs/7790936694/job/21245925157?pr=11869). This is the body of the failure:

```
=================================== FAILURES ===================================
___________________ test_cancel_scope_repr[AlarmCancelScope] ___________________
[gw7] linux -- Python 3.8.18 /opt/hostedtoolcache/Python/3.8.18/x64/bin/python

cls = <class 'prefect._internal.concurrency.cancellation.AlarmCancelScope'>

    @pytest.mark.parametrize(
        "cls", [AlarmCancelScope, WatcherThreadCancelScope, AsyncCancelScope]
    )
    async def test_cancel_scope_repr(cls):
        scope = cls()
        assert "PENDING" in repr(scope)
        assert "runtime" not in repr(scope)
        assert hex(id(scope)) in repr(scope)
    
>       with scope:

tests/_internal/concurrency/test_cancellation.py:47: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <AlarmCancelScopeat 0x7fee7e7427f0 RUNNING, runtime=0.02>

    def __enter__(self):
        super().__enter__()
    
        current_thread = threading.current_thread()
        self._previous_timer = None
    
        if current_thread is not threading.main_thread():
>           raise ValueError(
                "Alarm based timeouts can only be used in the main thread."
            )
E           ValueError: Alarm based timeouts can only be used in the main thread.

src/prefect/_internal/concurrency/cancellation.py:330: ValueError
```

--- a/tests/_internal/concurrency/test_cancellation.py
+++ b/tests/_internal/concurrency/test_cancellation.py
@@ -35,9 +35,40 @@ def mock_alarm_signal_handler():
         signal.signal(signal.SIGALRM, _previous_alarm_handler)
 
 
-@pytest.mark.parametrize(
-    "cls", [AlarmCancelScope, WatcherThreadCancelScope, AsyncCancelScope]
-)
+@pytest.mark.timeout(method="thread")  # alarm-based pytest-timeout will interfere
+async def test_alarm_cancel_scope_repr(cls=AlarmCancelScope):
+    scope = cls()
+    assert "PENDING" in repr(scope)
+    assert "runtime" not in repr(scope)
+    assert hex(id(scope)) in repr(scope)
+
+    if threading.current_thread() is threading.main_thread():
+        with scope:
+            assert "RUNNING" in repr(scope)
+            assert "runtime" in repr(scope)
+
+        assert "COMPLETED" in repr(scope)
+        assert "runtime" in repr(scope)
+
+    if threading.current_thread() is threading.main_thread():
+        scope = cls()
+        try:
+            with scope:
+                scope.cancel()
+        except CancelledError:
+            pass
+
+        assert "CANCELLED" in repr(scope)
+
+    scope = cls(name="test")
+    assert hex(id(scope)) not in repr(scope)
+    assert "name='test'" in repr(scope)
+
+    scope = cls(timeout=0.1)
+    assert "timeout=0.1" in repr(scope)
+
+
+@pytest.mark.parametrize("cls", [WatcherThreadCancelScope, AsyncCancelScope])
 async def test_cancel_scope_repr(cls):
     scope = cls()
     assert "PENDING" in repr(scope)

--- a/tests/_internal/concurrency/test_cancellation.py
+++ b/tests/_internal/concurrency/test_cancellation.py
@@ -35,9 +35,8 @@ def mock_alarm_signal_handler():
         signal.signal(signal.SIGALRM, _previous_alarm_handler)
 
 
-@pytest.mark.timeout(method="thread")  # alarm-based pytest-timeout will interfere
-async def test_alarm_cancel_scope_repr(cls=AlarmCancelScope):
-    scope = cls()
+async def test_alarm_cancel_scope_repr():
+    scope = AlarmCancelScope()
     assert "PENDING" in repr(scope)
     assert "runtime" not in repr(scope)
     assert hex(id(scope)) in repr(scope)
@@ -51,7 +50,7 @@ async def test_alarm_cancel_scope_repr(cls=AlarmCancelScope):
         assert "runtime" in repr(scope)
 
     if threading.current_thread() is threading.main_thread():
-        scope = cls()
+        scope = AlarmCancelScope()
         try:
             with scope:
                 scope.cancel()
@@ -60,11 +59,11 @@ async def test_alarm_cancel_scope_repr(cls=AlarmCancelScope):
 
         assert "CANCELLED" in repr(scope)
 
-    scope = cls(name="test")
+    scope = AlarmCancelScope(name="test")
     assert hex(id(scope)) not in repr(scope)
     assert "name='test'" in repr(scope)
 
-    scope = cls(timeout=0.1)
+    scope = AlarmCancelScope(timeout=0.1)
     assert "timeout=0.1" in repr(scope)
 
 

